compiler/rustc_mir/src/transform/simplify_try.rs

```diff
@@ -690,7 +691,13 @@ impl<'a, 'tcx> SimplifyBranchSameOptimizationFinder<'a, 'tcx> {
 
             match rhs {
                 Rvalue::Use(operand) if operand.place() == Some(adt_matched_on) => {
-                    StatementEquality::ConsideredEqual(side_to_choose)
+                    // FIXME(76803): This logic is currently broken because it does not take into
+                    // account the current discriminant value.
+                    if mir_opt_level > 2 {
```

Comment:
So by asking for opt-level 3 I still get the buggy optimization?
What is the status of the plans to make it less easy to accidentally get incorrect optimizations?

Reply:
The status is that we've made a decision but haven't implemented it yet

Reply:
I'm almost ready to push a PR with that change but I didn't want to do it here so that the beta backport would be as minimal as possible. I'll be sure to tag you @RalfJung when I open that PR.

Reply:
To be clear, this is the PR that @wesleywiser is referring to: https://github.com/rust-lang/rust/pull/76899